### PR TITLE
Implement kwargs in function calls

### PR DESCRIPTION
**Pull request by OddBloke**

### CODE REVIEW

File: parser/src/ast.rs [hunk 1]
```diff
@@ -137,7 +137,8 @@ pub enum Expression {
     },
     Call {
         function: Box<Expression>,
-        args: Vec<Expression>,
+        // parameters are (None, value), kwargs are (keyword name, value)
+        args: Vec<(Option<String>, Expression)>,
```
Comment (windelbouwman) [inline parser/src/ast.rs]: Are kwargs allowed before normal args? If not, I would suggest two vectors, args and kwargs. In any case I would like to have a seperate struct for kwargs with a name and expression member.

**Comment by windelbouwman** [comment]
This would be cool, so we could implement `int("17", base=7)`!

File: vm/src/vm.rs [hunk 2]
```diff
@@ -821,8 +868,37 @@ impl VirtualMachine {
...
+            }
+            bytecode::Instruction::CallFunctionKw { count } => {
+                let kwarg_names = self.pop_value();
+                let args: Vec<PyObjectRef> = self.pop_multiple(*count);
+
+                let kwarg_names = kwarg_names
+                    .to_vec()
+                    .unwrap()
+                    .iter()
+                    .map(|pyobj| pyobj.to_str().unwrap())
```
Comment (windelbouwman) [inline vm/src/vm.rs]: vm.to_str might be the better option here?

File: vm/src/pyobject.rs [hunk 3]
```diff
@@ -512,21 +512,30 @@ impl fmt::Debug for PyObject {
 #[derive(Debug, Default, Clone)]
 pub struct PyFuncArgs {
     pub args: Vec<PyObjectRef>,
-    // TODO: add kwargs here
+    pub kwargs: Option<Vec<(String, PyObjectRef)>>,
```
Comment (windelbouwman) [inline vm/src/pyobject.rs]: Finally :+1:

Comment (windelbouwman) [reply vm/src/pyobject.rs]: Btw, is option required? An empty vector means no kwargs?

File: parser/src/ast.rs (hunk 1 above)
Comment (OddBloke) [reply parser/src/ast.rs]: They aren't, but these are created in the lalrpop-generated code. I couldn't work out a good way of expressing the ordering there, so at the moment, we enforce the ordering in the compiler.

The problem is, I think, that with `foo(x, y="z")` the parser can't know when it examines `y` whether or not it is a positional argument or a keyword argument.

My grammar-fu is very weak though, so there may well be a way of handling that!

File: vm/src/pyobject.rs (hunk 3 above)
Comment (OddBloke) [reply vm/src/pyobject.rs]: Nope, not required; just pushed up a fix.

File: vm/src/vm.rs (hunk 2 above)
Comment (OddBloke) [reply vm/src/vm.rs]: These are the keyword argument names, which can _only_ be literal strings:

```
$ python -c 'foo("s"+"a"="foo")'
  File "<string>", line 1
SyntaxError: keyword can't be an expression
```

so we shouldn't ever get in to a case where this isn't a string; I've just pushed up a commit to use `objstr::get_value` directly.

File: parser/src/ast.rs (hunk 1 above)
Comment (windelbouwman) [reply parser/src/ast.rs]: Okay, I'll have a look later on.

**Pull request merged**